store/primary/primary.go

```diff
@@ -109,55 +173,53 @@ func (s *rtStorage) put(k string, entry store.IndexEntry) bool {
 		return false
 	}
 
-	s.current.Put(k, append(old, entry))
+	c.current.Put(k, append(old, entry))
 	return true
 }
 
-func (s *rtStorage) rotateCache() {
-	// Acquire all locks
-	for i := range s.branchLocks {
-		s.branchLocks[i].Lock()
-	}
-
-	if s.current.Len() >= s.sizeLimit {
-		s.previous, s.current = s.current, radixtree.New()
-	}
-
-	// Release all locks
-	for i := range s.branchLocks {
-		s.branchLocks[i].Unlock()
+func (c *syncCache) remove(k string, entry store.IndexEntry) bool {
+	removed := removeEntry(c.current, k, entry)
+	if removeEntry(c.previous, k, entry) {
+		removed = true
 	}
+	return removed
 }
 
-// PutMany puts store.IndexEntry information in several CIDs.
-// This is usually triggered when a bulk update for a providerID-pieceID
-// arrives.
-func (s *rtStorage) PutMany(cids []cid.Cid, provID peer.ID, pieceID cid.Cid) error {
-	in := store.IndexEntry{ProvID: provID, PieceID: pieceID}
-	var addedEntry bool
-
-	for i := range cids {
-		k := string(cids[i].Hash())
-		s.lockBranch(k)
+func removeEntry(cache *radixtree.Bytes, k string, entry store.IndexEntry) bool {
+	// Get from current cache
+	v, found := cache.Get(k)
+	if !found {
+		return false
+	}
 
-		if s.put(k, in) {
-			addedEntry = true
+	values := v.([]store.IndexEntry)
+	for i := range values {
+		if entry.PieceID == values[i].PieceID &&
```

Comment:
Reading this I just realized something. I used a list of tuples for `IndexEntry` because I assumed several providers wouldn't be storing the same CID, and by using a list I prevented having to loop through a map to return every value (I recall adding a NOTE somewhere). However, this assumption may be wrong as CIDs may be representing small blocks that are indexed as part of several larger pieces of content. This may be adding an unnecessary overhead by repeating the providerID in every tuple of the list. 

Again, I don't have real data to support this claim, but just wanted to share it here in case you feel it is worth adding a comment somewhere or filing an issue.

Reply:
I suppose all situations are possible:
1.  CID is provided by one provider, but with multiple piece IDs
2. CID is provided by multiple providers, with same piece IDs
3. CID is provided by multiple providers, with different piece IDs

If we need to support all of these, then a CID will need to map to a list of (providerID, pieceID) pairs.